File: PR8.1/Additional/huffman.py
from heapq import heappush, heappop, heapify
from collections import defaultdict, Counter

# Cấu trúc Node của cây Huffman
class Node:
    def __init__(self, ch, freq):
        self.ch = ch
        self.freq = freq
        self.left = None
        self.right = None
    
    # Để so sánh các Node trong hàng đợi ưu tiên
    def __lt__(self, other):
        return self.freq < other.freq
    
# Hàm tạo mã Huffman
def encode(root, prefix, huffman_code):
    if root is None:
        return
    
    if root.left is None and root.right is None:
        huffman_code[root.ch] = prefix

    encode(root.left, prefix + '0', huffman_code)
    encode(root.right, prefix + '1', huffman_code)

# Hàm giải mã chuỗi đã mã hóa
def decode(root, index, encoded_str):
    if root is None:
        return index, None
    
    if root.left is None and root.right is None:
        return index, root.ch
    
    index += 1
    if encoded_str[index] == '0':
        return decode(root.left, index, encoded_str)
    else:
        return decode(root.right, index, encoded_str)
    
# Xây dựng cây Huffman và mã hóa văn bản
def build_huffman_tree(text):
    # Tính tần số của các kí tự
    freq = Counter(text)

    # Tạo hàng đợi ưu tiên và chèn từng kí tự với tần số của nó
    heap = [Node(ch, freq) for ch, freq in freq.items()]
    heapify(heap)

    # Xây dựng cây Huffman
    while len(heap) > 1:
        left = heappop(heap)
        right = heappop(heap)
        merged = Node(Node, left.freq + right.freq)
        merged.left = left
        merged.right = right
        heappush(heap, merged)

    root = heap[0]

    # Tạo mã Huffman cho từng kí tự
    huffman_code = {}
    encode(root, '', huffman_code)

    print("Huffman Codes:")
    for ch, code in huffman_code.items():
        print(f"{ch}: {code}")

    print("\nHuffman algorithm")

    print("----------------")
    print("Original string:\n" + text)
    print("Size in bits:", len(text) * 8)

    # Nén văn bản
    encoded_str = ''.join(huffman_code[ch] for ch in text)
    print("\nEncoded string:\n" + encoded_str)
    print("Size in bits:", len(encoded_str))

    # Giải mã văn bản
    print("\nDecoded string:")
    index = -1
    while index < len(encoded_str) - 1:
        index, ch = decode(root, index, encoded_str)
        print(ch, end='')

    print("\n----------------\n")

# Phần chính của chương trình
text = "Suni"
while text != "-1":
    text = input("Write the string (write -1 to exit): ")

    if text != "-1":
        build_huffman_tree(text)

print("Program exited")

File: PR8.1/Additional/huffman.cpp
#include <iostream>
#include <unordered_map>
#include <queue>
#include <vector>
#include <string>
#include <fstream>

using namespace std;

// Cấu trúc cho một node trong cây Huffman
struct Node {
    char ch;
    int freq;
    Node *left, *right;

    Node(char ch, int freq, Node* left = nullptr, Node* right = nullptr) {
        this->ch = ch;
        this->freq = freq;
        this->left = left;
        this->right = right;
    }
};

// So sánh để sắp xếp trong hàng đợi đầu tiên
struct Compare {
    bool operator()(Node* l, Node* r) {
        return l->freq > r->freq;
    }
};

// Duyệt cây Huffman và mã hóa các kí tự
void encode(Node* root, string str, unordered_map<char, string> &huffmanCode) {
    if (root == nullptr)
        return;

    if (!root->left && !root->right) {
        huffmanCode[root->ch] = str;
    }

    encode(root->left, str + "0", huffmanCode);
    encode(root->right, str + "1", huffmanCode);
}

// Duyệt cây Huffman để giải mã
void decode(Node* root, int &index, string str) {
    if (root == nullptr)
        return;

    if (!root->left && !root->right) {
        cout << root->ch;
        return;
    }

    index++;

    if (str[index] == '0')
        decode(root->left, index, str);
    else
        decode(root->right, index, str);
}

// Hàm để xây dựng cây Huffman và thực hiện mã hóa
void buildHuffmanTree(string text) {
    unordered_map<char, int> freq;
    for (char ch : text) {
        freq[ch]++;
    }

    priority_queue<Node*, vector<Node*>, Compare> pq;

    for (auto pair : freq) {
        pq.push(new Node(pair.first, pair.second));
    }

    while (pq.size() != 1) {
        Node *left = pq.top(); pq.pop();
        Node *right = pq.top(); pq.pop();

        int sum = left->freq + right->freq;
        pq.push(new Node('\0', sum, left, right));
    }

    Node* root = pq.top();

    unordered_map<char, string> huffmanCode;
    encode(root, "", huffmanCode);

    cout << "Huffman Codes:\n";
    for (auto pair : huffmanCode) {
        cout << pair.first << ": " << pair.second << endl;
    }

    cout << "\nHuffman algorithm" << endl;

    cout << "----------------" << endl;
    cout << "Original string:\n" << text << endl;
    cout << "Size in bits: " << text.length() * 8 << endl;

    string str = "";
    for (char ch : text) {
        str += huffmanCode[ch];
    }

    cout << "\nEncoded string:\n" << str << endl;
    cout << "Size in bits: " << str.length() << endl;

    int index = -1;
    cout << "\nDecoded string:\n";
    while (index < (int)str.size() - 1) {
        decode(root, index, str);
    }
    cout << "\n----------------\n" << endl;
}

int main() {
    string data = "Suni";

    while (data != "-1") {
        cout << "Write the string (write -1 to exit): ";
        getline(cin, data);

        if (data != "-1") {
            buildHuffmanTree(data);
        }
    }

    cout << "Program exited" << endl;
    return 0;
}
